File: demo-app/app/src/javaApp/java/com/launchkey/android/authenticator/demo/ui/activity/AuthRequestActivityViewModel.java
package com.launchkey.android.authenticator.demo.ui.activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;

import com.launchkey.android.authenticator.sdk.core.auth_request_management.AuthRequest;
import com.launchkey.android.authenticator.sdk.core.auth_request_management.AuthRequestManager;
import com.launchkey.android.authenticator.sdk.core.auth_request_management.AuthRequestResponse;
import com.launchkey.android.authenticator.sdk.core.auth_request_management.event_callback.AuthRequestResponseEvent;
import com.launchkey.android.authenticator.sdk.core.auth_request_management.event_callback.GetAuthRequestEventCallback;
import com.launchkey.android.authenticator.sdk.core.util.CompositeDisposable;
import com.launchkey.android.authenticator.sdk.core.util.Disposable;

import java.util.concurrent.Executor;

public class AuthRequestActivityViewModel extends ViewModel {

    private final AuthRequestManager authRequestManager;
    private final Executor executor;

    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    public AuthRequest currentAuthRequest = null;
    private MutableLiveData<State> state = new MutableLiveData<>();

    private final GetAuthRequestEventCallback getAuthRequestEventCallback = new GetAuthRequestEventCallback() {
        @Override
        public void onSuccess(@Nullable final AuthRequest authRequest) {
            if (authRequest != null && currentAuthRequest != null && authRequest.getId().equals(currentAuthRequest.getId())) {
                state.setValue(new State.Success(currentAuthRequest));
                return;
            }

            currentAuthRequest = authRequest;
            state.setValue(new State.Success(currentAuthRequest));
        }

        @Override
        public void onFailure(@NonNull final Exception e) {
            state.setValue(new State.Failed(e));
        }
    };
    private final AuthRequestResponseEvent authRequestResponseEvent = new AuthRequestResponseEvent() {
        @Override
        public void onSuccess(@NonNull final AuthRequestResponse result) {
            currentAuthRequest = null;
            state.setValue(new State.Success(null));
        }

        @Override
        public void onFailure(@NonNull final Exception e) {
            currentAuthRequest = null;
            state.setValue(new State.Failed(e));
        }
    };

    public AuthRequestActivityViewModel(@NonNull final AuthRequestManager authRequestManager,
                                        @NonNull final Executor executor,
                                        @Nullable final SavedStateHandle savedStateHandle) {
        this.authRequestManager = authRequestManager;
        this.executor = executor;

        this.authRequestManager.registerForEvents(getAuthRequestEventCallback, authRequestResponseEvent);
    }

    public LiveData<State> getState() {
        return state;
    }

    @Override
    protected void onCleared() {
        compositeDisposable.clear();
        authRequestManager.unregisterForEvents(getAuthRequestEventCallback, authRequestResponseEvent);
    }

    public void checkForAuthRequest() {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                final Disposable d = authRequestManager.checkForAuthRequest(null);
                state.postValue(new State.Loading(d));
                compositeDisposable.add(d);
            }
        });
    }

    public static abstract class State {
        public static class Success extends State {
            @Nullable
            private final AuthRequest authRequest;
            Success(@Nullable final AuthRequest authRequest) {
                this.authRequest = authRequest;
            }

            @Nullable
            public AuthRequest getAuthRequest() {
                return authRequest;
            }
        }

        public static class Loading extends State {
            @NonNull
            private final Disposable disposable;
            Loading(@NonNull final Disposable disposable) {
                this.disposable = disposable;
            }

            @NonNull
            public Disposable getDisposable() {
                return disposable;
            }
        }

        public static class Failed extends State {
            @NonNull
            private final Exception exception;
            Failed(@NonNull final Exception exception) {
                this.exception = exception;
            }

            @NonNull
            public Exception getException() {
                return exception;
            }
        }
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/activity/AuthRequestActivityViewModel.kt
package com.launchkey.android.authenticator.demo.ui.activity

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import com.launchkey.android.authenticator.sdk.core.auth_request_management.AuthRequest
import com.launchkey.android.authenticator.sdk.core.auth_request_management.AuthRequestManager
import com.launchkey.android.authenticator.sdk.core.auth_request_management.AuthRequestResponse
import com.launchkey.android.authenticator.sdk.core.auth_request_management.event_callback.AuthRequestResponseEvent
import com.launchkey.android.authenticator.sdk.core.auth_request_management.event_callback.GetAuthRequestEventCallback
import com.launchkey.android.authenticator.sdk.core.util.CompositeDisposable
import com.launchkey.android.authenticator.sdk.core.util.Disposable
import java.util.concurrent.Executor

class AuthRequestActivityViewModel(private val authRequestManager: AuthRequestManager,
                                   private val executor: Executor,
                                   savedStateHandle: SavedStateHandle?) : ViewModel() {
    private val compositeDisposable = CompositeDisposable()
    var currentAuthRequest: AuthRequest? = null
    private val state = MutableLiveData<State>()
    private val getAuthRequestEventCallback: GetAuthRequestEventCallback = object : GetAuthRequestEventCallback() {
        override fun onSuccess(authRequest: AuthRequest?) {
            if (authRequest != null && currentAuthRequest != null && authRequest.id == currentAuthRequest!!.id) {
                state.value = State.Success(currentAuthRequest)
                return
            }
            currentAuthRequest = authRequest
            state.value = State.Success(currentAuthRequest)
        }

        override fun onFailure(e: Exception) {
            state.value = State.Failed(e)
        }
    }
    private val authRequestResponseEvent: AuthRequestResponseEvent = object : AuthRequestResponseEvent() {
        override fun onSuccess(result: AuthRequestResponse) {
            currentAuthRequest = null
            state.value = State.Success(null)
        }

        override fun onFailure(e: Exception) {
            currentAuthRequest = null
            state.value = State.Failed(e)
        }
    }

    fun getState(): LiveData<State> {
        return state
    }

    override fun onCleared() {
        compositeDisposable.clear()
        authRequestManager.unregisterForEvents(getAuthRequestEventCallback, authRequestResponseEvent)
    }

    fun checkForAuthRequest() {
        executor.execute {
            val d = authRequestManager.checkForAuthRequest(null)
            state.postValue(State.Loading(d))
            compositeDisposable.add(d)
        }
    }

    abstract class State {
        class Success internal constructor(val authRequest: AuthRequest?) : State()

        class Loading internal constructor(val disposable: Disposable) : State()

        class Failed internal constructor(val exception: Exception) : State()
    }

    init {
        authRequestManager.registerForEvents(getAuthRequestEventCallback, authRequestResponseEvent)
    }
}